lib/spack/spack/solver/asp.py

```diff
@@ -835,9 +833,10 @@ def package_compiler_defaults(self, pkg):
         ppk = spack.package_prefs.PackagePrefs(pkg.name, 'compiler', all=False)
         matches = sorted(compiler_list, key=ppk)
 
-        for i, cspec in enumerate(matches):
+        for i, cspec in enumerate(reversed(matches)):
             self.gen.fact(fn.node_compiler_preference(
-                pkg.name, cspec.name, cspec.version, i))
+                pkg.name, cspec.name, cspec.version, -i * 100
+            ))
```

Comment:
@tgamblin This is a bit of magic I had to do to make the concretizer pass a test on compiler preferences. The gist of it is that with multiple compilers we need to ensure that the weight of a preference for a single package is "negative enough" to be selected for minimization. Not sure this is the best formula though. Maybe:
```
- (i + 1) * 100
```
would be better as a weight, or maybe we should just change the target of minimization.